cumulusci/tasks/bulkdata/delete.py

```diff
@@ -38,77 +43,61 @@ def _init_options(self, kwargs):
         for obj in self.options["objects"]:
-            self.logger.info(f"Deleting {self._object_description(obj)} ")
-            delete_job = self._create_job(obj, self.options["where"])
-            if delete_job is not None:
-                self._wait_for_job(delete_job, error_behaviour="raise")
+            query = f"SELECT Id FROM {obj}"
+            if self.options["where"]:
+                query += f" WHERE {self.options['where']}"
 
-    def _create_job(self, obj, where=None):
-        # Query for rows to delete
-        delete_rows = self._query_salesforce_for_records_to_delete(obj, where)
-        if not delete_rows:
-            self.logger.info(f"  No {obj} objects found, skipping delete")
-            return
-
-        # Upload all the batches
-        operation = "hardDelete" if self.options["hardDelete"] else "delete"
-        delete_job = self.bulk.create_job(obj, operation)
-        self.logger.info(f"  Deleting {len(delete_rows)} {obj} records")
-        batch_num = 1
-        for batch in self._upload_batches(delete_job, delete_rows):
-            self.logger.info(f"    Uploaded batch {batch}")
-            batch_num += 1
-        self.bulk.close_job(delete_job)
-        return delete_job
+            self.logger.info(f"Querying for {obj} objects")
+            qs = BulkApiQueryOperation(
+                sobject=obj, api_options={}, context=self, query=query
+            )
+            qs.query()
+            if qs.job_result.status is not DataOperationStatus.SUCCESS:
+                raise BulkDataException(
+                    f"Unable to query records for {obj}: {','.join(qs.job_result.job_errors)}"
+                )
 
-    def compose_query(self, obj, where):
-        query = f"SELECT Id FROM {obj}"
-        if where:
-            query += f" WHERE {where}"
+            if not qs.job_result.total_records:
+                self.logger.info("No records found, skipping delete operation")
+                return
 
-        return query
+            self.logger.info(f"Deleting {self._object_description(obj)} ")
+            ds = BulkApiDmlOperation(
+                sobject=obj,
+                operation=(
+                    DataOperationType.HARD_DELETE
+                    if self.options["hardDelete"]
+                    else DataOperationType.DELETE
+                ),
+                api_options={},
+                context=self,
+                fields=["Id"],
+            )
+            ds.start()
+            ds.load_records(qs.get_results())
+            ds.end()
+
+            if ds.job_result.status not in [
+                DataOperationStatus.SUCCESS,
+                DataOperationStatus.ROW_FAILURE,
+            ]:
+                raise BulkDataException(
+                    f"Unable to delete records for {obj}: {','.join(qs.job_result.job_errors)}"
+                )
+
+            for result in ds.get_results():
+                if not result.success and not self.options["ignore_row_errors"]:
```

Comment:
Can we include the text for the failure reason in this (at least one row)?

Arguably we could also save the results to a file and let them peruse them themselves, but we'd get 90% of the benefit if we could see the first row.

Reply:
Yep. Changed this to parallel how it already works in `load.py`.